Lib/test/pickletester.py

```diff
@@ -2047,36 +2048,43 @@ def check_frame_opcodes(self, pickled):
         framed by default and are therefore considered a frame by themselves in
         the following consistency check.
         """
-        last_arg = last_pos = last_frame_opcode_size = None
-        frameless_opcode_sizes = {
-            'BINBYTES': 5,
-            'BINUNICODE': 5,
-            'BINBYTES8': 9,
-            'BINUNICODE8': 9,
-        }
+        frame_end = frameless_start = None
+        frameless_opcodes = {'BINBYTES', 'BINUNICODE', 'BINBYTES8', 'BINUNICODE8'}
         for op, arg, pos in pickletools.genops(pickled):
-            if op.name in frameless_opcode_sizes:
-                if len(arg) > self.FRAME_SIZE_TARGET:
-                    frame_opcode_size = frameless_opcode_sizes[op.name]
-                    arg = len(arg)
+            if frame_end is not None:
+                if op.name == 'FRAME':
+                    self.assertEqual(pos, frame_end)
                 else:
-                    continue
-            elif op.name == 'FRAME':
-                frame_opcode_size = 9
-            else:
-                continue
-
-            if last_pos is not None:
-                # The previous frame's size should be equal to the number
-                # of bytes up to the current frame.
-                frame_size = pos - last_pos - last_frame_opcode_size
-                self.assertEqual(frame_size, last_arg)
-            last_arg, last_pos = arg, pos
-            last_frame_opcode_size = frame_opcode_size
-        # The last frame's size should be equal to the number of bytes up
-        # to the pickle's end.
-        frame_size = len(pickled) - last_pos - last_frame_opcode_size
-        self.assertEqual(frame_size, last_arg)
+                    self.assertLessEqual(pos, frame_end)
+                self.assertLessEqual(pos, frame_end)
+                if pos < frame_end:
+                    self.assertNotIn(op.name, 'FRAME')
+                    if op.name in frameless_opcodes:
+                        self.assertLessEqual(len(arg), self.FRAME_SIZE_TARGET)
```

Comment:
Could you add a comment explaining this?

Reply:
Done.